### PR TITLE
Valarray loops expansion

### LINKED ISSUE #991: <valarray>: Avoid control flow in macro
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/stl/inc/valarray#L1094-L1124

Consider converting to a function that takes lambda or manually re-using the code.
Motivation: easier debugging (step by step or crash dump).

Occurrences:
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/stl/inc/valarray#L60-L72
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/stl/inc/valarray#L874-L879
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/stl/inc/valarray#L1014-L1020
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/stl/inc/valarray#L1094-L1102
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/stl/inc/valarray#L1192-L1197

### LINKED ISSUE #1169: <valarray>: magic static
**Describe the bug**
valarray potentially uses magic static

**Command-line test case**
```
#include <string>
#include <iostream>
#include <valarray>

int main()
{
	std::valarray<std::string> a{"strings", "are", "numeric", "types"};
	std::valarray<std::string> b = a.shift(2);
	for (std::string s : b) {
		std::cout << s << " ";
	}
}
```

**Expected behavior**
It works as expected. 
But in C++17 where `std::string` constructor is not `constexpr` it should rely on magic static in debug mode:
https://github.com/microsoft/STL/blob/e000d3feaf6acaf6214aedcfa21de483931ccbb7/stl/inc/valarray#L353

**STL version**
https://github.com/microsoft/STL/commit/e000d3feaf6acaf6214aedcfa21de483931ccbb7

**Additional context**
Probably not a real problem, only from lawyering point of view.

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Resolves #991,  resolves #1169.

Note that `_MOP` macro was not expanded directly as is, it was simplified to reduce number of lines.

### CODE REVIEW

File: stl/inc/valarray [hunk 1]
```diff
@@ -195,99 +184,175 @@ public:
     valarray& operator=(const indirect_array<_Ty>& _Indarr); // defined below
 
     _NODISCARD valarray operator+() const {
-        _VALOP(_Ty, size(), +_Myptr[_Idx]);
+        valarray<_Ty> _Ans(size());
```
Comment (miscco) [inline stl/inc/valarray:187]: Should we capture `size()` so that we do not need the call to `_Ans.size((`

File: stl/inc/valarray [hunk 2]
```diff
@@ -195,99 +184,175 @@ public:
     valarray& operator=(const indirect_array<_Ty>& _Indarr); // defined below
 
     _NODISCARD valarray operator+() const {
-        _VALOP(_Ty, size(), +_Myptr[_Idx]);
+        valarray<_Ty> _Ans(size());
+        for (size_t _Idx = 0; _Idx < _Ans.size(); ++_Idx) {
+            _Ans[_Idx] = +_Myptr[_Idx];
```
Comment (miscco) [inline stl/inc/valarray:189]: I believe the `+` is unnecessary

Comment (AlexGuteniev) [reply stl/inc/valarray:189]: Are you sure that for any `NumericType` unary + is no operation?

File: stl/inc/valarray (hunk 1 above)
Comment (AlexGuteniev) [reply stl/inc/valarray:187]: I think as `_Ans` is a local variable, compiler can optimize `size()` to not reading from memory address by itself

Comment (miscco) [reply stl/inc/valarray:187]: But debug  performance. 

I am totally fine with it, given that it is existing code

Comment (AlexGuteniev) [reply stl/inc/valarray:187]: Right, debug will be affected...

File: stl/inc/valarray (hunk 2 above)
Comment (miscco) [reply stl/inc/valarray:189]: Urgh, the only thing I could imagine was integer promotion, but that should not be any problem as we asign the numeric type anyway and there cannot be any overflow.

**Comment by AlexGuteniev** [comment]
Nothing special can happen with normal integers or floats.

Assume some weird type where unary + does some normalization. Lambda is quite close example, although its + causes type conversion.

File: stl/inc/valarray (hunk 2 above)
Comment (AlexGuteniev) [reply stl/inc/valarray:189]: Nothing special can happen with normal integers or floats.

Assume some weird type where unary + does some normalization. Lambda is quite close example, although its + causes type conversion.

Comment (AlexGuteniev) [reply stl/inc/valarray:189]: And Numeric Type is defined by `valarray` quite broad. `std::string` is a Numeric Type, for example

File: stl/inc/valarray (hunk 1 above)
Comment (AlexGuteniev) [reply stl/inc/valarray]: Done.
(If maintainers disagree, I'll revert)

File: stl/inc/valarray [hunk 3]
```diff
@@ -351,37 +474,56 @@ public:
 
     _NODISCARD valarray shift(int _Count) const {
         __PURE_APPDOMAIN_GLOBAL static _Ty _Dflt;
```
Comment (AlexGuteniev) [inline stl/inc/valarray:476]: Reported as #1169 , can fix here too

File: stl/inc/valarray [hunk 4]
```diff
@@ -476,373 +598,743 @@ _NODISCARD const _Ty* end(const valarray<_Ty>& _Array) {
...
+    return _Ans;
 }
 
 template <class _Ty>
 _NODISCARD _Boolarray operator&&(const typename valarray<_Ty>::value_type& _Left, const valarray<_Ty>& _Right) {
-    _VALOP(bool, _Right.size(), _Left&& _Right[_Idx]);
+    const size_t _Size = _Right.size();
+    valarray<bool> _Ans(_Size);
+    for (size_t _Idx = 0; _Idx < _Size; ++_Idx) {
+        _Ans[_Idx] = _Left && _Right[_Idx];
```
Comment (StephanTLavavej) [inline stl/inc/valarray:814]: Bonus clang-format improvement! :smile_cat: (No change requested.)

File: stl/inc/valarray [hunk 5]
```diff
@@ -983,8 +1504,8 @@ public:
 
     size_t _Off(_Sizarray& _Indexarr) const {
         size_t _Idx, _Ans = _Start;
```
Comment (StephanTLavavej) [inline stl/inc/valarray:1506]: Note to self: there will be a trivial adjacent-line merge conflict with #1226 here. No change requested.

**Comment by StephanTLavavej** [comment]
Thanks for making this code clearer, improving debug codegen, and finding and fixing the magic static that eluded us! :smiley_cat:

**Pull request merged**